src/tpm/highlevel.rs

```diff
@@ -0,0 +1,304 @@
+where
+    F: Fn(&mut Context) -> Result<T>,
+{
+    let oldses = ctx.sessions();
+    ctx.set_sessions(ses);
+
+    let res = f(ctx);
+
+    ctx.set_sessions(oldses);
+
+    res
+}
+
+fn execute_with_session<F, T>(
+    ctx: &mut Context,
+    ses: Option<Session>,
+    f: F,
+) -> Result<T>
+where
+    F: Fn(&mut Context) -> Result<T>,
+{
+    execute_with_sessions(ctx, (ses, None, None), f)
+}
+
+fn execute_with_temp_nullauth_session<F, T>(
+    ctx: &mut Context,
+    f: F,
+) -> Result<T>
+where
+    F: Fn(&mut Context) -> Result<T>,
+{
+    let ses = create_nullauth_session(ctx)?.unwrap();
+
+    let res = execute_with_session(ctx, Some(ses), f);
+
+    ctx.flush_context(ses.handle().into())?;
+
+    res
+}
+
+fn execute_without_session<F, T>(ctx: &mut Context, f: F) -> Result<T>
+where
+    F: Fn(&mut Context) -> Result<T>,
+{
+    execute_with_session(ctx, None, f)
+}
+
+pub(crate) fn retrieve_ek_pubcert(
+    ctx: &mut Context,
+    alg: AsymmetricAlgorithm,
+) -> Result<Vec<u8>> {
+    let nv_idx = match alg {
+        AsymmetricAlgorithm::Rsa => RSA_2048_EK_CERTIFICATE_NV_INDEX,
+        AsymmetricAlgorithm::Ecc => ECC_P256_EK_CERTIFICATE_NV_INDEX,
+    };
+
+    let nv_idx = NvIndexTpmHandle::new(nv_idx).unwrap();
+    let nv_idx = TpmHandle::NvIndex(nv_idx);
+    let nv_idx = ctx.tr_from_tpm_public(nv_idx)?;
+    let nv_idx: NvIndexHandle = nv_idx.into();
+
+    let (nvpub, _) = execute_without_session(ctx, |ctx| {
+        ctx.nv_read_public(nv_idx).map_err(|e| e.into())
+    })?;
+    let nvsize = nvpub.data_size();
+
+    execute_with_temp_nullauth_session(ctx, |ctx| {
+        nv_read_full(ctx, nv_idx.into(), nv_idx, nvsize)
+    })
+}
+
+// Source: TCG EK Credential Profile for TPM Family 2.0; Level 0 Version 2.3 Revision 2
+// Appendix B.3.3 and B.3.4
+fn create_ek_public_from_default_template(
+    alg: AsymmetricAlgorithm,
+) -> Result<TPM2B_PUBLIC> {
+    let mut obj_attrs = ObjectAttributes(0);
+    obj_attrs.set_fixed_tpm(true);
+    obj_attrs.set_st_clear(false);
+    obj_attrs.set_fixed_parent(true);
+    obj_attrs.set_sensitive_data_origin(true);
+    obj_attrs.set_user_with_auth(false);
+    obj_attrs.set_admin_with_policy(true);
+    obj_attrs.set_no_da(false);
+    obj_attrs.set_encrypted_duplication(false);
+    obj_attrs.set_restricted(true);
+    obj_attrs.set_decrypt(true);
+    obj_attrs.set_sign_encrypt(false);
+
+    // TPM2_PolicySecret(TPM_RH_ENDORSEMENT)
+    // With 32 null-bytes attached, because of the type of with_auth_policy
+    let authpolicy: [u8; 64] = [
+        0x83, 0x71, 0x97, 0x67, 0x44, 0x84, 0xb3, 0xf8, 0x1a, 0x90, 0xcc,
+        0x8d, 0x46, 0xa5, 0xd7, 0x24, 0xfd, 0x52, 0xd7, 0x6e, 0x06, 0x52,
+        0x0b, 0x64, 0xf2, 0xa1, 0xda, 0x1b, 0x33, 0x14, 0x69, 0xaa, 0x00,
+        0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00,
+        0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00,
+        0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00,
+    ];
```

Comment:
Is this likely to remain static, or would anyone want to ever change this? I am wondering is this should be scooped out of a config file instead of coded in here?

Reply:
This is actually a statically defined value from the specification.